flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleMaster.java

```diff
@@ -41,4 +44,17 @@
 	CompletableFuture<T> registerPartitionWithProducer(
 		PartitionDescriptor partitionDescriptor,
 		ProducerDescriptor producerDescriptor);
+
+	/**
+	 * Release any external resources occupied by the given partition.
+	 *
+	 * <p>This call triggers release of any resources which are occupied by the given partition in the external systems
+	 * outside of the producer executor. This is mostly relevant for the batch jobs and blocking result partitions
+	 * for which {@link ResultPartitionDeploymentDescriptor#isReleasedOnConsumption()} returns {@code false}.
+	 * The producer local resources are managed by {@link ShuffleDescriptor#hasLocalResources()} and
+	 * {@link ShuffleEnvironment#releasePartitions(Collection)}.
+	 *
+	 * @param shuffleDescriptor shuffle descriptor of the result partition to release externally.
+	 */
+	void releasePartitionExternally(T shuffleDescriptor);
```

Comment:
I'm not sure if this signature will work due to generics.

For this to work one has to maintain the generic shuffle descriptor type from `registerPartitionWithProducer` until `releasePartitionExternally`. As it stands this isn't possible. The shuffle master in the EG is currently defined as `ShuffleMaster<?> shuffleMaster`.

Reply:
An alternative could be to simply accept a `ResultPartitionID` instead. This would force some book-keeping into the implementations, but i don't see a solution right now.

Reply:
True, we have to switch to `ShuffleDescriptor` type there because we do not keep static type information about concrete `ShuffleDescriptor` in JM.

Reply:
Having generified `ShuffleMaster` is under question then, we can consider removing generics later..